invenio_rdm_records/metadata_extensions.py

```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2019 CERN.
+# Copyright (C) 2019 Northwestern University.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""Metadata Extensions."""
+from copy import deepcopy
+
+from flask import current_app
+from invenio_records_rest.schemas.fields import DateString, SanitizedUnicode
+from marshmallow import Schema
+from marshmallow.fields import Bool, Integer, List
+
+
+class MetadataExtensions(object):
+    """Custom metadata extensions helper class."""
+
+    def __init__(self, namespaces, extensions):
+        """Constructor.
+
+        :param namespaces: Namespaces dict
+
+        See :const:`invenio_rdm_records.config.RDM_RECORDS_METADATA_NAMESPACES`
+        Example:
+
+        .. code-block:: python
+
+                {
+                    'dwc': {
+                        '@context': 'http://rs.tdwg.org/dwc/terms/'
+                    },
+                    'z': {
+                        '@context': 'https://zenodo.org/terms'
+                    }
+                }
+
+        :param extensions: Extensions dict
+
+        See :const:`invenio_rdm_records.config.RDM_RECORDS_METADATA_EXTENSIONS`
+        Example:
+
+        .. code-block :: python
+
+                {
+                    'dwc:family': {
+                        'elasticsearch': 'keyword',
+                        'marshmallow': SanitizedUnicode()
+                    },
+                    'dwc:behavior': {
+                        'elasticsearch': 'text',
+                        'marshmallow': SanitizedUnicode()
+                    },
+                    'z:department': {
+                        'elasticsearch': 'text',
+                        'marshmallow': SanitizedUnicode()
+                    }
+                }
+
+
+        Note: use translation layer `_()` to get human readable label from
+              ids used internally.
+        """
+        self.namespaces = deepcopy(namespaces) or {}
+        self.extensions = deepcopy(extensions) or {}
+        self._validate()
+
+    def _validate_marshmallow_type(self, field_cfg):
+        """Make sure the Marshmallow type is one we support."""
+        def validate_basic_marshmallow_type(_type):
+            allowed_types = [
+                Bool, DateString, Integer, SanitizedUnicode
+            ]
+            assert any([
+                isinstance(_type, allowed_type) for allowed_type
```

Comment:
Shelved: What about subclasses of any of the types?

Reply:
I was thinking we should be conservative at first: only accept very clear fields and over time see if the needs for more refined fields does come about. To revisit when the need does show.